clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/Types.hs

```diff
@@ -0,0 +1,161 @@
+-- | Contains the types used in the implementation of the GMII to SGMII bridge wrapper.
+-- LogiCORE IP Ethernet 1000BASE-X PCS/PMA or SGMII
+
+{-# LANGUAGE LambdaCase #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=10 #-}
+{-# LANGUAGE RecordWildCards #-}
+module Clash.Cores.Xilinx.Ethernet.Gmii.Types where
+
+import Clash.Prelude
+
+data Lvds = Lvds{lvdsP :: Bit, lvdsN :: Bit}
+  deriving (Generic, NFDataX, BitPack)
+
+-- TODO: Incorporate in 'Clash.Prelude'?
+-- | Gets time in 'Picoseconds' from time in 'Nanoseconds'.
+type Nanoseconds  (ns :: Nat) = Picoseconds  (1000 * ns)
+-- | Gets time in 'Picoseconds' from time in 'Picoseconds', essentially 'id'.
+type Picoseconds  (ps :: Nat) = ps
+-- | A generic GMII interface
+
+-- TODO: Move to re-usable module?
+-- | Represents a GMII interface.
+data Gmii = Gmii
+ { gmiiData :: BitVector 8
+ , gmiiValid :: Bit
+ , gmiiError :: Bit
+ } deriving (Generic, BitPack, NFDataX)
+
+data DuplexMode = HalfDuplex | FullDuplex
+  deriving (Generic, NFDataX, BitPack, Eq)
+
+-- | Link speeds in Mega bits per second.
+data LinkSpeed = Speed10 | Speed100 | Speed1000
+  deriving (Generic, NFDataX, BitPack)
+
+-- | Whether the link partner supports pause frames.
+data Pause = NoPause | SymmetricPause | AsymmetricPause | SymmetricAsymmetricPause
+  deriving (Generic, NFDataX, BitPack)
+
+type AutoNegConfigVector = BitVector 16
+-- | A subset of the an_adv_config_vector that is applicable for the GMII to SGMII bridge
+data AutoNegConfig =
+  AutoNegConfig
+  { cLinkSpeed :: LinkSpeed
+  , cDuplexMode :: DuplexMode
+  } deriving (Generic, NFDataX, BitPack)
+
+-- | Converts the AutoNegConfig Records to the an_adv_config_vector format.
+toAutoNegConfigVector :: AutoNegConfig -> AutoNegConfigVector
+toAutoNegConfigVector AutoNegConfig{..} = bv4
+  where
+    (linkSpeed0, linkSpeed1) = case cLinkSpeed of
+      Speed10 -> (0, 0)
+      Speed100 -> (0, 1)
+      Speed1000 -> (1, 0)
```

Comment:
It feels like this could just be done with `pack @LinkSpeed` when the constructors are ordered suitably.

Reply:
The Auto negotiation configuration vector and status vector both contain link speed information, but their formats differ.

I'd like to not rely implicitly on the order of the constructors for the correct behavior.

Reply:
Ah, that is a shame. I think it's best to make the `BitPack` match one of those and use explicit conversion for the other (see elsewhere for why I still want to use `BitPack`). That way you at least save some LUTs.